pallets/deitos/src/lib.rs

```diff
@@ -113,54 +135,180 @@ pub mod pallet {
+            status: IPStatus,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            let bounding_amount = IPBoundingAmount::<T>::get()?;
+            let ip = T::Lookup::lookup(provider)?;
+
+            InfrastructureProvider::<T>::try_mutate(
+                &ip,
+                |ip_details| -> Result<_, DispatchError> {
+                    let ip_details = ip_details
+                        .as_mut()
+                        .ok_or(Error::<T>::NonExistentStorageValue)?;
+
+                    ip_details.status = status.clone();
+
+                    Self::deposit_event(Event::IPStatusChanged {
+                        ip: ip.clone(),
+                        status,
+                    });
+
+                    Ok(())
+                },
+            )?;
 
             Ok(())
         }
 
         #[pallet::call_index(2)]
+        #[pallet::weight(T::WeightInfo::update_ip_storage())]
+        pub fn update_ip_storage(
+            origin: OriginFor<T>,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        ) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer.
+            let ip = ensure_signed(origin)?;
+
+            InfrastructureProvider::<T>::try_mutate(
+                &ip,
+                |ip_details| -> Result<_, DispatchError> {
+                    let ip_details = ip_details
+                        .as_mut()
+                        .ok_or(Error::<T>::NonExistentStorageValue)?;
+
+                    // Check if the new total_storage is enough to cover the current reserved_storage
+                    ensure!(
+                        total_storage >= ip_details.reserved_storage,
+                        Error::<T>::InsufficientStorage
+                    );
+
+                    ip_details.price_storage_per_block = price_storage_per_block;
+                    ip_details.total_storage = total_storage;
+                    // Note: This calculation might need to be reviewed or adjusted.
+                    ip_details.reserved_storage = ip_details
+                        .total_storage
+                        .saturating_sub(ip_details.reserved_storage);
+
+                    Self::deposit_event(Event::IPStorageUpdated {
+                        ip: ip.clone(),
+                        price_storage_per_block,
+                        total_storage,
+                    });
+
+                    Ok(())
+                },
+            )?;
+
+            Ok(())
+        }
+
+        #[pallet::call_index(3)]
         #[pallet::weight(T::WeightInfo::shutdown_ip())]
         pub fn shutdown_ip(origin: OriginFor<T>, price_unit: BalanceOf<T>) -> DispatchResult {
             // Check that the extrinsic was signed and get the signer.
-            let who = ensure_signed(origin)?;
+            let ip = ensure_signed(origin)?;
+
+            InfrastructureProvider::<T>::try_mutate(
```

Comment:
Doing this you prevent an IP from comming back. Is there a reason for doing this instead of simply removing the guy from state?

Reply:
we cannot remove the guy from the state i think for easily querying historical purposes. Also i suspect for the onchain reputation system we might need to have the existing storage item.

Reply:
yes, makes a lot of sense. don't we need a transaction that allows an IP to enable himself back?

Reply:
I have adapted the register_ip function to handle both cases (new IP or unregistered IP that wants to join back):

```rust
            // Checks that the IP is either not registered or is registered but with Unregistered status
            if let Some(ip_details) = InfrastructureProvider::<T>::get(&ip) {
                ensure!(
                    ip_details.status == IPStatus::Unregistered,
                    Error::<T>::IPAlreadyExists
                );
            }
```